daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala

```diff
@@ -272,6 +277,10 @@ abstract class SequenceParserBase(
               // the sequence succesfully
               isDone = true
               pstate.setSuccess()
+              // If we're masking the failure, we don't want the error dianostics
+              // to flow up. Restore the diagnostics from before the parse
+              // attempt
+              pstate.diagnostics = diagnosticsBeforeAttempt
```

Comment:
I've added an additional test that I think shows that discriminated unordered sequences are just broken, even before these changes. I discovered this when trying to figure out why there's a single test failure with these changes.

I think the issue is that unordred sequences are implemented by repeatedly parsing a ChoiceParser. When the ChoiceParser fails to parse a branch, then the sequence parser assume we are done with with the unordered sequence. When that happens, we hit this failure case, throw away that failure and the "all alternatives failed" diagnostics, and continue parsing like normal. The ChoiceParser will have reset us back to the the PoU, so things just work.

However, say we have a discriminated unordred sequence (e.g. initiated content). Now let's assume an initiator was found. This will resolve the PoU created in the ChoiceParser. Now say there is a failue parsing that discriminated branch. The PoU was resolved so we will not attempt any alternatives of the unordered sequence/ChoiceParser and we immediately create a parse error. We then return back to here, and subsequently ignore that error and associated diagnostics. And because the PoU was resolved, we don't even backtrack and will just continue parsing whereever the choice branch left off.

The new test shows this behavior. Even though an assertion is created, it is ignored and the parse finishes. This is incorrect.

I'm not sure yet what the right fix is, but I think things are just totally broken with discriminated unordred sequences. It feels to me like maybe we can't reuse the ChoiceParser. I think the sequence parser needs to be the one to create the PoU associated with each alternative of the unordered sequence rather than relying on the ChoiceParser. That way it can know if the PoU was resolved or not, and then correctly end the sequence or propogate an error upwards. Right now, the ChoiceParser does everything so we don't have enough information in the SequeneParser to act correctly.

Note that the SequenceParser does have some logic around UnorderedDiscriminatedFailure, but I think that's just broken. The case above on line 273 is never actually hit in codecov. I'm guessing that this stuff broke when we change how PoU's work.

Thoughts?

Reply:
One thought is we should split this into two issues. 

One is the fact that diagnostics are being created and retained despite backtracking, and that one is important to fix. 

The other is a bug with unordered sequences, and the importance of that may not be as high - depends on whether there are DFDL Schemas depending on it and are broken by this. 

VCard is the only published DFDL schema I can find that uses unordered sequences.

An unordered sequence is a lot like an array of choice, but yeah, my guess based on your note above is we can't directly reuse choice parser, but must make a variant of it (sharing traits/bases perhaps) that has somewhat different behavior for discriminators since logically an unordered sequence does not have two nested PoUs, just one per child element instance.

Reply:
I'll do a regression test and see if there are any failures with these changes. If not, I agree that it's probably worth merging this and open a new issue for discriminated unordred sequences. I think that combination is rare enough that it's unlikely that any schemas use the feature.

Reply:
All my regression tests pass except for vCard, which has one test that runs into this exact issue. vCard has this exact issue--unordered sequences with initated content

So if a unordered sequence branch fails in vCard, we lose the diagnostic about why that initiated content branch failed. We will then continue trying to parse after the unorderd sequence fails (which is is this second pre-existing bug). In this case, we run into another parse error. Because we threw away the error we don't get the old helpful error message.

So this fixes the bug related to incorrectly showing diagnostics, at the expense of sometimes not showing diagnostics, and sometimes not even failing a parse when we should. I'll create a bug for this new issue. Seems pretty important to fix, but considering it only affects this one case, I'm not sure it's a blocker.

Reply:
Created [DAFFODIL-2512](https://issues.apache.org/jira/browse/DAFFODIL-2512) for this issue.